deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -833,6 +850,50 @@ def increment_value(dictionary, key):
     def overlapping_partition_gradients_reduce_epilogue(self):
         self.independent_gradient_partition_epilogue()
 
+    def update_separate_grad_accum(self):
+        for group in self.bit16_groups:
+            for param in group:
+                if param.grad is not None:
+                    if param.grad_accum is None:
+                        param.grad_accum = param.grad.to(
+                            self.gradient_accumulation_dtype)
+                    else:
+                        param.grad_accum.add_(
+                            param.grad.to(self.gradient_accumulation_dtype).view(
+                                param.grad_accum.shape))
+                    # Don't clear if you still need the native grad for reduction (stage 2)
+                    if self.use_grad_accum_for_reduction:
+                        param.grad = None
+
+    def set_grad_accum_pointer(self):
+        for group in self.bit16_groups:
+            for param in group:
+                param.grad_accum = param.grad
+
+    # Clear the tensor the reduction gradient attribute is pointing to
+    def clear_grad_reduc_pointer(self, param):
+        param.grad_reduc = None
+        if self.use_grad_accum_for_reduction:
+            param.grad_accum = None
+        else:
+            param.grad = None
+
+    # Point to the gradient tensor to use in reduction
+    def set_grad_reduc_pointer(self, param):
```

Comment:
I think we can avoid this setter by using a getter to access the gradient to reduce. Something like 

```python
  def get_gradient_for_reduction(self, param):
    return param.grad_accum if self.use_grad_accum_for_reduction else param.grad
```

Then all uses of `param.grad_reduc` can be replaced.